mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -1,31 +1,473 @@
+        };
+        (key, topic.clone())
+    }
+
+    pub async fn start(&self) -> Result<(), BridgeError> {
         info!("Starting bridge...{}", self.connection_settings.name());
 
-        self.connect_to_local().await;
-        self.connect_to_remote().await;
+        self.connect_to_local().await?;
+        self.connect_to_remote().await?;
+
+        Ok(())
+    }
+
+    async fn connect_to_remote(&self) -> Result<(), BridgeError> {
+        info!(
+            "connecting to remote broker {}",
+            self.connection_settings.address()
+        );
+
+        self.connect(
+            &self.subscriptions,
+            self.connection_settings.address(),
+            *self.connection_settings.keep_alive(),
+            self.connection_settings.clean_session(),
+            self.connection_settings.credentials(),
+        )
+        .await
+    }
+
+    async fn connect_to_local(&self) -> Result<(), BridgeError> {
+        let client_id = format!(
+            "{}/$edgeHub/$bridge/{}",
+            self.device_id,
+            self.connection_settings.name()
+        );
+        info!(
+            "connecting to local broker {}, clientid {}",
+            self.system_address, client_id
+        );
+
+        self.connect(
+            &self.forwards,
+            self.system_address.as_str(),
+            *self.connection_settings.keep_alive(),
+            self.connection_settings.clean_session(),
+            &Credentials::Anonymous(client_id),
+        )
+        .await
+    }
+
+    async fn connect(
+        &self,
+        topics: &HashMap<String, Topic>,
+        address: &str,
+        keep_alive: Duration,
+        clean_session: bool,
+        credentials: &Credentials,
+    ) -> Result<(), BridgeError> {
+        let mut topic_filters = Vec::new();
+        for val in topics.values() {
+            topic_filters.push(TopicMapper {
+                topic_settings: val.clone(),
+                topic_filter: TopicFilter::from_str(val.pattern())
+                    .map_err(BridgeError::TopicFilterParseError)?,
+            });
+        }
+
+        let mut client = MqttClient::new(
+            address,
+            keep_alive,
+            clean_session,
+            MessageHandler::new(topic_filters, BATCH_SIZE),
+            credentials,
+        );
+
+        let subscriptions: Vec<String> = topics.keys().map(|key| key.into()).collect();
+        debug!("subscribe to remote {:?}", subscriptions);
+
+        client
+            .subscribe(subscriptions)
+            .await
+            .map_err(BridgeError::SubscribeError)?;
+
+        //TODO: handle this with shutdown
+        let _events_task = tokio::spawn(client.handle_events());
+
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+struct TopicMapper {
+    topic_settings: Topic,
+    topic_filter: TopicFilter,
+}
+
+/// Handle events from client and saves them with the forward topic
+#[derive(Clone)]
+struct MessageHandler<'a, T>
+where
+    T: Persist<'a>,
```

Comment:
I believe we don't need lifetime and phantom altogether

Reply:
with the change that Andrew has in PR for persistence this will me removed.